synctool/src/main/java/org/duracloud/sync/endpoint/DuraStoreChunkSyncEndpoint.java

```diff
@@ -193,16 +198,81 @@ private ChunksManifest getManifest(String spaceId, String contentId) {
+
+        //clean up any orphaned chunks and / or obsolete unchunked files
+        try {
+            log.debug("checking for chunks, manifests, and/or unchunked files that should be removed : for {}/{}",
+                      spaceId, contentId);
+            Iterator<String> chunkedContentIdIt = store.getSpaceContents(getSpaceId(), contentId + ".dura-");
+            Set<String> chunkedContentIds = new HashSet<>();
+            chunkedContentIdIt.forEachRemaining(id -> chunkedContentIds.add(id));
+            //if there are chunks
+            if (!chunkedContentIds.isEmpty()) {
+                //clean up
+                if (syncFile.length() <= chunkerOptions.getMaxChunkSize()) {
+                    log.info("A chunked version was replaced by an unchunked version of {}/{}",
+                              spaceId, contentId);
+
+                    //if file is less than or equal to max chunk size
+                    //then we can expect the new file is unchunked
+                    //look for any chunked content matching path
+                    //delete all.
+                    chunkedContentIds.stream().forEach(content -> {
+                        deleteContent(spaceId, content, store);
+                    });
+                    log.info("Deleted manifest and all chunks associated with {}/{}",
+                              spaceId, contentId);
+                } else {
+                    log.info("Checking for orphaned chunks associated with {}/{}",
+                             spaceId, contentId);
+
+                    //resolve the set of the chunks in the manifest
+                    ChunksManifest manifest = getManifest(spaceId, contentId);
+                    Set<String> manifestChunks = new HashSet<>();
+                    manifest.getEntries().stream().forEach(entry -> manifestChunks.add(entry.getChunkId()));
+                    //for each chunk in storage, delete if not in the manifest.
+                    chunkedContentIds.stream()
+                                     .filter(chunkedContentId -> !chunkedContentId.endsWith(manifestSuffix))
+                                     .forEach(chunk -> {
+                                         if (!manifestChunks.contains(chunk)) {
+                                             log.debug("Chunk not found in manifest: deleting orphaned chunk ({}/{})",
+                                                       spaceId, chunk);
+                                             deleteContent(spaceId, chunk, store);
+                                         }
+                                     });
+                    //check for an unchunked version and remove it if happens to exist.
+                    if (store.contentExists(spaceId, contentId)) {
+                        deleteContent(spaceId, contentId, store);
+                    }
+                }
+            }
+        } catch (Exception ex) {
+            throw new RuntimeException(ex);
```

Comment:
Not sure if throwing here is the best option. If the file transfer completed properly, then addUpdateContent() wouldn't be called again, so this chunk check code wouldn't be retried. If somehow addUpdateContent() were called again, then the file would be transferred again even though that appears to have completed properly. Maybe it's best just to log an ERROR and leave it be?

Reply:
My thought here was that in the (probably rare) case of an error in the clean up process, throwing an error would cause the file to be retried (and cleaned up again).  If the file had been transferred successfully, the synctool would recalculate checksums and check them against storage, but it would not retransfer the content, no?  So there would be some duplicated effort on the one hand, and yet throwing an exception would ensure that things are properly cleaned up.

Reply:
You're right that the RuntimeException would cause the file to be retried, but it's what happens in that retry that is problematic. Looking at DuraStoreSyncEndpoint.syncFileAndReturnDetailedResult(), if jumpstart is on, the file will be transferred again (no checksum check) and chunk cleanup will run again. If jumpstart is not on, then a checksum check occurs. If the checksum of the local file matches what's in DuraCloud (which is most likely if the failure was during chunk cleanup), then addUpdateContent() does not get called, meaning chunk cleanup isn't run.

So the result of the retry is either an unnecessary re-transfer of the file or an unnecessary checksum check. Only in one of those two results in chunk cleanup being run again. 

Perhaps it would be better to pull the try/catch block here out into its own method and add some retry direct logic rather than relying on the file retry mechanism?

Reply:
@bbranan :  ah okay.  I get it.  I will address that.